Add task to download files from url

safern opened the pull request:
This task will be used in order to download CoreCLR symbol package in CoreFX. 

cc: @weshaggard @danmosemsft

danmosemsft commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+        /// The directory where the file will be downloaded to.
+        /// </summary>
+        [Required]
+        public string DestinationDirectory { get; set; }
```
If I have to specify both anyway, why not just have Destination that takes a path? Taht way if I have a path, I don't have to split it to pass it in.

safern replied:
Makes sense. I did it to make sure that the directory existed before creating the file, but I can achieve that by getting the parent directory from the full path, and create it before creating the file. (This is because File.Create throws if part of the path doesn't exist)

danmosemsft commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+        /// The default is to emit warnings and not fail, the caller can fail if output parameter DownloadedFile is empty.
+        /// Set this to true to fail instead of just warn.
+        /// </summary>
+        public bool FailOnError { get; set; }
```
maybe TreatErrorsAsWarnings, it is used elsewhere

danmosemsft commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+{
+    public sealed class DownloadFileFromUrl : BuildTask
+    {
+        
```
nit.. line

danmosemsft commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+                        {
+                            if (result.IsSuccessStatusCode)
+                            {
+                                var bytes = result.Content.ReadAsByteArrayAsync().GetAwaiter().GetResult();
```
Can you rewrite this to write to the file async? And avoid the intermediate byte array with ReadAsStreamAsync.

I realize that perf may not be critical but it's nice to do it the "right" way as it's no more lines.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+                {
+                    using (var client = new HttpClient())
+                    {
+                        using (var result = client.GetAsync(DownloadSource).GetAwaiter().GetResult())
```
please include some logic for handling the proxy. See https://github.com/dotnet/corefx/pull/23467 as example.

safern replied:
I've fixed this. Talked to @davidsh and there is a difference on how this is done in .NET Core vs Desktop. The way to do it in .NET Core is not available until .NET4.7.1 so I had to use if def.

mellinoe commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,92 @@
+        /// The file that was downloaded, if there is an error this will be empty.
+        /// </summary>
+        [Output]
+        public string DownloadedFile { get; set; }
```
If we combine the two "destination" input parameters into "DestinationPath" (for example), then this will just be equivalent to that.

safern commented:
Addressed PR Feedback. I've added a cleanup at the end to delete the file created if the download failed in order to track it in the targets to see if we need to unzip the file or not. Instead I could add an Output param but I'll wait for feedback.

danmosemsft commented:
You were going to make it async?

danmosemsft commented:
Oh, nm, I see your comment

safern commented:
> You were going to make it async?

We need to wait for the file to be created because we are going to unzip it and then extract the contents from another task. I’m using CopyToFileAsync but I have to wait for it anyway so that we can get the contents in the target.

danmosemsft commented:
You can wait for the file to finish writing, but it should be possible for it to write while the download completes.

joperezr commented on src/Microsoft.DotNet.Build.Tasks/DownloadFileFromUrl.cs:
```diff
@@ -0,0 +1,108 @@
+            bool success = false;
+            try
+            {
+                Log.LogMessage(MessageImportance.High, $"Downloading {DownloadSource} -> {DestinationFile}");
```
NIT: do we really want this to be `MessageImportance.High`? I would expect this type of messages to not be present on minimal logging.

safern replied:
I made it like this because downloading a file can take ~5 or ~10 seconds, so that when we are building we know we are downloading something, but I can remove it and leave it to the targets to log that.

joperezr replied:
that is true for this specific use because it's downloading a nuget package, but this task can/will be reused to download other things in the future, and I'm not sure if we will always want to have the verbosity in all cases. That's why I would suggest to keep logging a message, just not make it Importance.High

safern replied:
Sounds reasonable. We could also pass an optional parameter to enable High verbosity and log with Importance.High if that is enabled.

weshaggard replied:
I wouldn't do that. Just make the logging normal verbosity and if folks want that information they can always enable that level of logging.

safern replied:
Ok, sounds good. Will do that then.

joperezr commented:
I'm wondering if it would be useful to add also a .targets file called `DownloadSymbols.targets` or something and have that do all the logic of given an itemGroup of sources (defined by each repo) it will do the download packages, mark them as zip files, extract them, and get the .pdbs out. I say this becuase it seems that this logic could potentially be shared across different repos, and as you can see there are several steps that we might not want to have duplicated. Just of the top of my head, WCF jumps to mind, since they will probably want to do this too.

joperezr commented:
If we do agree to add the targets file though, that doesn't need to block this PR, it could be done on a subsequent PR.

safern commented:
I've changed it to write to the file while downloading instead of first download and then write to the file, and also to be able to receive multiple items to download and custom metadata per item like `DestinationDir`. 

Could you guys please take a look?

weshaggard commented on src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs:
```diff
@@ -0,0 +1,144 @@
+                    }
+                }
+
+                string fileName = item.GetMetadata("DestinationFile");
```
don't we need to set this if it isn't found? Wouldn't we default to the filename of the requested file.

safern replied:
But we actually don't have the filename of the requested file, as it is a url and the url doesn't always in the filename that we want to download i.e to download a symbol package it is `https://feedurl/symbols/packageId/packageVersion` so we don't really have a way to know the default filename.

safern replied:
There will be some cases were the Url will have the filename at the end, but that is not always the case, that is why I wanted to force people to give the destination file name (which they should always know).

weshaggard replied:
What do you download in that case?

safern replied:
Whatever the HTTP GET request returns, which in the case of the symbol server it returns a symbols.nupkg file. If the request doesn't return anything or throws an error then we would just output a a warning or an error, depending on the `TreatErrorsAsWarnings` parameter.

weshaggard replied:
OK perhaps default the filename based on the url if it has a filename.

safern replied:
OK if we can't find a filename in the URL and the DestinationFile metadata is empty then we throw the error.

weshaggard replied:
Yeah otherwise we have an empty filename.

safern replied:
But empty file name will cause an exception when trying to create the file, do we want to handle that error ourselves or should we leave it to `File.OpenWrite` ?

weshaggard replied:
I think we should handle it ourselves so we can give a better error message.

safern replied:
OK.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs:
```diff
@@ -0,0 +1,174 @@
+                    // starts with file://
+                    if (downloadUri.IsFile)
+                    {
+                        fileName = Path.GetFileName(downloadUri.LocalPath);
```
I think you can actually use Path.GetFileName on both cases instead of trying to parse the Uri yourself below.

safern replied:
Ok, will try that. In that case we wouldn't need the downloadUri.IsFile check. Will try it.

safern replied:
Fixed. Thanks for the feedback.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks/DownloadFilesFromUrl.cs:
```diff
@@ -0,0 +1,174 @@
+                        // with this we ensure that if the metadata doesn't contain a file name, we always get one.
+                        fileName = urlContents[urlContents.Length - 1];
+
+                        if (string.IsNullOrEmpty(fileName))
```
I would move this check one level up just to be sure that the file case also gets a filename that isn't empty.

safern commented:
Thanks for the feedback!
